### PR TITLE
[SPARK-32555][SQL] Add unique ID on query execution

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR adds unique ID on QueryExecution, so that listeners can leverage the ID to deduplicate redundant calls.

### Why are the changes needed?

I've observed that Spark calls QueryExecutionListener multiple times on same QueryExecution instance (even same funcName for onSuccess). There's no unique ID on QueryExecution, hence it's a bit tricky if the listener would like to deal with same query execution only once.

Note that streaming query has both query ID and run ID which can be leveraged as unique ID.

### Does this PR introduce _any_ user-facing change?

Yes for who uses query execution listener - they'll see `id` field in QueryExecution and leverage it.

### How was this patch tested?

Manually tested. I think the change is obvious hence don't think it warrants a new UT. StreamingQueryListener has been using UUID as `queryId` and `runId` so it should work for the same.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -54,6 +54,8 @@ class QueryExecution(
     val logical: LogicalPlan,
     val tracker: QueryPlanningTracker = new QueryPlanningTracker) extends Logging {
 
+  val id: UUID = UUID.randomUUID()
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:57]: Should we maybe follow who we do in `Dataset.id` or `SQLExecution.nextExecutionId`?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:57]: I guess you meant using monotonically incremental long value; yes that would also work. Will update.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
I also agree with this proposal. +1, LGTM.
Merged to master for Apache Spark 3.1.0 on December 2020.
Thank you, @HeartSaVioR and @HyukjinKwon .

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Comment by dongjoon-hyun** [comment]
Also, cc @dbtsai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -54,6 +55,8 @@ class QueryExecution(
     val logical: LogicalPlan,
     val tracker: QueryPlanningTracker = new QueryPlanningTracker) extends Logging {
 
+  val id: Long = QueryExecution.nextExecutionId
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:58]: What is the use case to add this ID?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:58]: We have a listener which parses the logical & physical plan from the query execution and reflects the changes to Apache Atlas. While it works well in general, there's a pain point on the listener due to the fact Spark can callback the listener multiple times on the same query execution, whereas the instance of query execution doesn't have the unique ID to distinct. That's already I wrote in PR description as well as JIRA description.

**Pull request closed without merging**